### DISCUSSION TITLE
Completion source - differences between lazydev and nvim_lsp?

### DISCUSSION THREAD

**Discussion by josh-nz** [folke/lazydev.nvim#23] [Q&A]
I can't see any difference between these two, but I've only tired the basics.

I currently have 
```Lua
 sources = cmp.config.sources({
        { name = "nvim_lsp", } ...
```
and everything (for my basic needs) looks to work still when using lazydev. When I add
```Lua
      sources = cmp.config.sources({
        { name = "nvim_lsp", },
        {
           name = "lazydev",
            group_index = 0, -- set group index to 0 to skip loading LuaLS completions
        },
```
Everything still works. I've tried typing `require("` without the lazydev completion source and it looks to offer me the full range of suggestions beyond the currently loaded files.

What are benefits of adding lazydev as a completion source vs just continuing to use nvim_lsp?

**Comment by folke** [comment]
`nvim_lsp` can only complete `require("...` calls for libraries added to your workspace.
Since lazydev dynamically only loads the libraries that are in use, you won't have the requires available before that.
So that's where the cmp source comes in

**Reply by josh-nz** [reply]
Right. Since I'm pretty much only editing Lua files in the Neovim workspace (my config files), and don't do any other sort of Lua programming, I don't think this completion source is required by me, if I understand correctly.

**Reply by folke** [reply]
If you only use it to edit your config, then you **do** need it. Or at least then it is very helpful.
lazydev is **only** useful for editing neovim config files.

But you of course also need `nvim_lsp`

**Reply by josh-nz** [reply]
Ok, playing around a little more I see what you describe. Completion for `require` only suggests my Neovim modules (loaded as part of the workspace automatically I'm guessing), but not plugins, unless a file I have open `require`s a plugin, then that plugin module is loaded into the LSP (by lazydev) and therefore `nvim_lsp` completion for that module becomes available.

However, even when I add the `lazydev` completion source, I'm not seeing the unloaded plugin modules. Update: I've been able to see the `lazydev` completion sources, and it does (sometimes) work. However, there are cases where it doesn't show this completion, when it should. I have a set of steps to reliably reproduce the behaviour. I'm unsure if this is in part due to my completion setup, a bug in the completion plugin, or a bug in the `lazydev` source. Where would you suggest I start with this? Should I log an issue here in the lazydev repo?

**Comment by Ezall2001** [comment]
Hello, I am having the same problem here no unloaded packages are shown in the completion. is there a 'lazydev-cmp' we need to install ?
